torchvision/transforms.py

```diff
@@ -380,3 +380,31 @@ def __call__(self, img):
         scale = Scale(self.size, interpolation=self.interpolation)
         crop = CenterCrop(self.size)
         return crop(scale(img))
+
+
+class Whiten(object):
```

Comment:
Can we add a test too that given a whitening matrix this transforms produces an image with Identity covariance?

Reply:
Also, given that we pass the whitening matrix as an argument, this could be renamed to have a more general name.
For example, this layer performs a linear transformation of the input. Maybe this could be renamed something like `LinearTransformation`, and mention whitening as one of the applications?

Reply:
I also thought about this while doing the implementation. Should we also change `whitening_matrix` to something else then?

I mean: 

```
class LinearTransformation(object):
    def __init__(self, general_name_for_matrix):
        self.general_name_for_matrix = general_name_for_matrix
```

Leaving `whitening_matrix` with `LinearTransformation` might be confusing maybe?

For example this could be used also for linear random projections of the input. Also I think all the possible applications should be mentioned in the README, but should the docstring be shorter? Or do you think it is better to put everything in the docstring too?

Let's figure this out and I will go on with the implementation

Reply:
Yes, I think we should change the `whitening_matrix` to something else.
And we will hopefully converge to having documentation being automatically generated from docstrings, so I'd say to focus on adding this information in the docstrings.

Reply:
Regarding the naming, I'm going with `transformation_matrix`. If you have better ideas, I'm open to suggestions.

I'm having problems with the test: I compute the whitening matrix on the whole dataset by

```
# fake data
x = np.random.randn(1000, 144)

# zero center data: compute and subtract avg of 
# each feature on the 1000 observations
avg = np.mean(x, axis=0)[np.newaxis, :]
x-=avg

# estimate covariance and apply SVD
cov = np.dot(x.T, x)/x.shape[0]
u, s, _ = np.linalg.svd(cov)
xrot = np.dot(x, u)

# compute principal components
epsilon = 1e-10
pc = np.dot(np.dot(u, np.diag(1./np.sqrt(s+epsilon))), u.T)

# CASE 1: this produces something close to the identity 144x144
x_white = np.dot(x, pc)
cov = np.cov(x_white)

# CASE 2: this does not
xw = np.dot(x[0][np.newaxis, :], pc)
print(np.dot(xw.T, xw))
```

I get covariance=identity matrix only when I apply it on the whole dataset, if I apply it only on one sample (the case of `transforms`) I get something else. 
The test is a translation in Torch of Case 2 and it is failing. Now, I looked at the Keras implementation and it is doing the same thing, so... Is this normal? Do you see some mistake?

Reply:
Hi @iacolippo,

I'm not sure this line is correct `pc = np.dot(np.dot(u, np.diag(1./np.sqrt(s+epsilon))), u.T)` I think what you want to do is: `transformation_matrix = u / np.sqrt(s + epsilon)`?

In any case, I think the reason you are getting non-Identity covariance matrices is because you need to set the `rowvar` param to `False` in the call to `np.cov`

Reply:
Hi @alykhantejani 

adding the dot product with `u.T` is the difference between PCA and ZCA whitening as far as I know: https://github.com/zellyn/deeplearning-class-2011/blob/master/ufldl/pca_2d/pca_2d.py

I think I found the bug (taking a dot product in wrong order). I will write the test, check carefully and then push it.